modules/README.md

```diff
@@ -1,209 +1,58 @@
-    "status" : "unknown",
-    "error_type" : "no_puppet_bin",
-    "error" : "Puppet executable '$puppet_bin' does not exist",
-    "exitcode" : -1
-}
-```
-
-### The Puppet agent is busy performing a run when the run action is called
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "agent_already_running",
-    "error" : "Puppet is already running",
-    "exitcode" : -1
-}
-```
-
-### The Puppet agent has been disabled on the host the run action was called on
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "agent_disabled",
-    "error" : "Puppet is disabled",
-    "exitcode" : -1
-}
-```
-
-### The run action was called with invalid JSON on stdin
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "invalid_json",
-    "error" : "Invalid json parsed on STDIN",
-    "exitcode" : -1
-}
-```
-
-### The Puppet run failed to start
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "agent_failed_to_start",
-    "error" : "Failed to start Puppet",
-    "exitcode" : -1
-}
-```
-
-### Puppet exited with a non-zero exitcode
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "agent_exit_non_zero",
-    "error" : "Puppet exited with a non 0 exitcode",
-    "exitcode" : $exitcode
-}
-```
-
-### last_run_report.yaml doesn't exist after a successful run
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "no_last_run_report",
-    "error" : "$last_run_report.yaml doesn't exist",
-    "exitcode" : $exitcode
-}
-```
-
-### last_run_report.yaml exists but isn't valid yaml
-
-
-```
-{
-    "kind" : "unknown",
-    "time" : "unknown",
-    "transaction_uuid" : "unknown",
-    "environment" : "unknown",
-    "status" : "unknown",
-    "error_type" : "invalid_last_run_report",
-    "error" : "$last_run_report.yaml isn't valid yaml",
-    "exitcode" : exitcode
-}
-```
+# PXP agent architecture
+
+map of name:module, possibly factory style from peg
+
+Module abstraction
+
+    class Module {
+        std::string name;
+        bool invoke(const CthunMessage& request, CthunMessage& response);
+        Schema actions;
+    }
+
+
+## External modules
+
+Load from libdir
+
+Will have 1 required action, metadata, takes no parameters, yields a
+json description of the actions it has.  Parameters subkey should be a
```

Comment:
I can't find any explanation of what's meant by the {{parameters subkey}} - did you mean the response to the metadata action?